Cesium.CodeGen/Ir/Expressions/OffsetOfExpression.cs

```diff
@@ -0,0 +1,49 @@
+using Cesium.CodeGen.Contexts;
+using Cesium.CodeGen.Extensions;
+using Cesium.CodeGen.Ir.Declarations;
+using Cesium.CodeGen.Ir.Types;
+using Mono.Cecil.Cil;
+
+namespace Cesium.CodeGen.Ir.Expressions;
+
+internal class OffsetOfExpression : IExpression
+{
+    private readonly StructType _resolvedType;
+    private readonly LocalDeclarationInfo _fieldType;
+
+    public OffsetOfExpression(StructType resolvedType, LocalDeclarationInfo fieldType)
+    {
+        _resolvedType = resolvedType;
+        _fieldType = fieldType;
+    }
+
+    public IExpression Lower(IDeclarationScope scope) => this;
+
+    public void EmitTo(IEmitScope scope)
+    {
+        var typeRef = _resolvedType.Resolve(scope.Context);
+        var fieldRef = typeRef.Resolve().Fields.Single(f => f.Name == _fieldType.Identifier);
+
+        VariableDefinition var;
+
+        if (scope.Method.Body.Variables.FirstOrDefault(v => v.VariableType.IsEqualTo(typeRef)) is
+            { } existingVar)
+        {
+            var = existingVar;
+        }
+        else
+        {
+            var = new VariableDefinition(typeRef);
+            scope.Method.Body.Variables.Add(var);
+        }
+
+        scope.AddInstruction(OpCodes.Ldloca, var);
+        scope.AddInstruction(OpCodes.Ldflda, fieldRef);
```

Comment:
I think this should be intrinsic which just calculate value of offset, no rely on runtime for that. but.... On the second thought, what if we have .NET type from external library which would like to use. Personally I would still have constant evaluation first, and when we implement interop with .NET we special case these types.

Reply:
I think there's no need to dig into implementation-defined details of the runtime, it's better to use what runtime already provides

Reply:
Cesium has their own rules. All struct firstly defined by Cesium and layout is controlled by Cesium